src/java/org/apache/cassandra/db/guardrails/DefaultGuardrail.java

```diff
@@ -0,0 +1,690 @@
+        // (the internal "state" will be null)
+        ClientWarn.instance.warn(fullMessage);
+        // Similarly, tracing will also ignore the message if we're not running tracing on the current thread.
+        Tracing.trace(fullMessage);
+        for (Guardrails.Listener listener : Guardrails.listeners)
+            listener.onWarningTriggered(name, redactedMessage);
+    }
+
+    protected void warn(String fullMessage)
+    {
+        warn(fullMessage, fullMessage);
+    }
+
+    protected void fail(String fullMessage, String redactedMessage)
+    {
+        if (!skipNotifying(false))
+        {
+            logger.error(fullMessage);
+            // Tracing will ignore the message if we're not running tracing on the current thread.
+            Tracing.trace(fullMessage);
+            for (Guardrails.Listener listener : Guardrails.listeners)
+                listener.onFailureTriggered(name, redactedMessage);
+        }
+
+        if (throwOnFailure)
+            throw new InvalidRequestException(fullMessage);
+    }
+
+    protected void fail(String message)
+    {
+        fail(message, message);
+    }
+
+    /**
+     * Do no throw {@link InvalidRequestException} if a guardrail failure is triggered.
+     * <p>
+     * Note: this method is not thread safe and should only be used during guardrail initialization
+     *
+     * @return current guardrail
+     */
+    @Override
+    public DefaultGuardrail setNoExceptionOnFailure()
+    {
+        this.throwOnFailure = false;
+        return this;
+    }
+
+    /**
+     * Note: this method is not thread safe and should only be used during guardrail initialization
+     *
+     * @param minNotifyIntervalInMs frequency of logging and triggering listener to avoid spamming,
+     *                              default 0 means always log and trigger listeners.
+     * @return current guardrail
+     */
+    @Override
+    public DefaultGuardrail setMinNotifyIntervalInMs(long minNotifyIntervalInMs)
+    {
+        assert minNotifyIntervalInMs >= 0;
+        this.minNotifyIntervalInMs = minNotifyIntervalInMs;
+        return this;
+    }
+
+    /**
+     * Reset last notify time to make sure it will notify downstream when {@link this#warn(String, String)}
+     * or {@link this#fail(String)} is called next time.
+     */
+    @VisibleForTesting
+    public void resetLastNotifyTime()
+    {
+        lastFailInMs = 0;
+        lastWarnInMs = 0;
+    }
+
+    /**
+     * @return {@code true} if guardrail should not log message and trigger listeners; otherwise, update
+     * {@link #lastWarnInMs} or {@link #lastFailInMs} respectively.
+     */
+    private boolean skipNotifying(boolean isWarn)
+    {
+        if (minNotifyIntervalInMs == 0)
+            return false;
+
+        long nowInMs = currentTimeMillis();
+        long timeElapsedInMs = nowInMs - (isWarn ? lastWarnInMs : lastFailInMs);
+
+        boolean skip = timeElapsedInMs < minNotifyIntervalInMs;
```

Comment:
is there a reason to skip warning?  I get using no spam logger to not flood our logs, but if users are doing bad things shouldn't they get notified each time?

Reply:
That's meant to be used with the guardrails checking disk usage, that can be quite overwhelming. We can always remove this part from the prototype and add it back when we add those guardrails if we still see the need, wdyt?

Reply:
> add it back when we add those guardrails 

+1

Reply:
I have just removed that part.